File: 1825-find-minimum-time-to-finish-all-jobs/find-minimum-time-to-finish-all-jobs.cpp
class Solution {
private:
    bool rec(int mid, vector<int> &jobs, vector<int> &w, int k, int n, int idx){
        if(idx == n) return true;
        for(int i=0; i<k; i++){
            if(w[i]+jobs[idx] <= mid){
                w[i] += jobs[idx];
                if(rec(mid, jobs, w, k, n, idx+1)) return true;
                w[i] -= jobs[idx];
            }
            if(w[i] == 0) break;
        }
        return false;
    }
    bool check(int mid, vector<int> &jobs, int k, int n){
        vector<int> w(k, 0);
        return rec(mid,jobs,w,k,n,0);
    }
public:
    int minimumTimeRequired(vector<int>& jobs, int k) {
        sort(jobs.rbegin(), jobs.rend());
        int lo = jobs[0];
        int hi = accumulate(jobs.begin(), jobs.end(), 0);
        while (lo <= hi) {
            int mid = lo+(hi-lo)/2;
            if (check(mid, jobs, k, jobs.size())) hi = mid-1;
            else lo = mid+1;
        }
        return lo;
    }
};

File: 1825-find-minimum-time-to-finish-all-jobs/find-minimum-time-to-finish-all-jobs.py
class Solution:
    def minimumTimeRequired(self, jobs: List[int], k: int) -> int:
        jobs.sort(reverse=True)
        workers = [0] * k
        self.ans = sum(jobs)

        def backtrack(index):
            if index == len(jobs):
                self.ans = min(self.ans, max(workers))
                return

            for i in range(k):
                if workers[i] + jobs[index] >= self.ans:
                    continue

                workers[i] += jobs[index]
                backtrack(index + 1)
                workers[i] -= jobs[index]

                if workers[i] == 0:
                    break

        backtrack(0)
        return self.ans